File: projects/spring-boot-restv2/demo/src/main/java/com/sourabh/demo/service/TransactionService.java
package com.sourabh.demo.service;

import com.sourabh.demo.exceptions.SomethingNotFoundException;
import com.sourabh.demo.iservice.ITransactionService;
import com.sourabh.demo.model.Transaction;
import com.sourabh.demo.repository.TransactionRepository;
import org.bson.Document;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class TransactionService implements ITransactionService {
    @Autowired
    TransactionRepository transactionRepository;



    public List<Transaction> getAll(){
        List<Transaction> transactions = transactionRepository.findAll();
        if(transactions.isEmpty())
            throw new SomethingNotFoundException("No Transactions yet!!");
        return transactions;
    }

    public Transaction getById(int id) {
        return transactionRepository.findById(id).orElseThrow(() -> new SomethingNotFoundException("No Transaction for the given id: " + id));
    }

    public Transaction addOrUpdate(Transaction transaction){
        return transactionRepository.save(transaction);
    }

    public Document getRoomWiseEarnings(){
        return transactionRepository.getRoomWiseEarnings().getRawResults();
    }
}

File: projects/spring-boot-rest-kt/hotel-reservation/src/main/kotlin/com/sourabh/hotelreservation/service/TransactionService.kt
package com.sourabh.hotelreservation.service

import com.sourabh.hotelreservation.exception.SomethingNotFoundException
import com.sourabh.hotelreservation.iservice.ITransactionService
import com.sourabh.hotelreservation.model.Transaction
import com.sourabh.hotelreservation.repository.TransactionRepository
import org.bson.Document
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class TransactionService(@Autowired var transactionRepository: TransactionRepository):ITransactionService {
    override fun getAll(): MutableList<Transaction> {
        val transactions = transactionRepository.findAll()
        if(transactions.isEmpty())
            throw SomethingNotFoundException("No transactions yet!!")
        return transactions
    }

    override fun getById(id: Int): Transaction {
        return transactionRepository.findById(id).orElseThrow { SomethingNotFoundException("No Transaction found for the id: $id") }
    }

    override fun addOrUpdate(transaction: Transaction): Transaction {
        return transactionRepository.save(transaction)

    }
    fun getRoomWiseEarnings(): Document? {
        return transactionRepository.getRoomWiseEarnings()!!.rawResults
    }
}